ambry-commons/src/main/java/com/github/ambry/compression/BaseCompression.java

```diff
@@ -0,0 +1,290 @@
+   * The size of sourceDataBuffer must be at least the original data source size.
+   * The decompression implementation is algorithm specific.
+   *
+   * @param compressedBuffer The compressed buffer.
+   * @param compressedBufferOffset The offset in compressedBuffer where the decompression should start reading.
+   * @param compressedBufferSize Size of the compressed buffer returned from compressData().
+   * @param sourceDataBuffer The buffer to store decompression output (the original source data).
+   * @param sourceDataOffset Offset where to write the decompressed data.
+   * @param sourceDataSize Size of the buffer to hold the decompressed data.  It should be size of original data.
+   */
+  protected abstract void decompressData(byte[] compressedBuffer, int compressedBufferOffset, int compressedBufferSize,
+      byte[] sourceDataBuffer, int sourceDataOffset, int sourceDataSize);
+
+  /**
+   * Get the algorithm name in binary format.
+   * This binary will be saved into the compressed buffer.  The algorithm name will be used to find the decompressor.
+   * The byte array contains the size (1 byte) and the algorithm name.
+   * The binary generation happens only once per algorithm (cached) and is thread safe.
+   *
+   * @return Binary presentation of algorithm name.
+   */
+  private byte[] getAlgorithmNameBinary() {
+    if (algorithmNameBinary != null) return algorithmNameBinary;
+
+    synchronized(this) {
+      if (algorithmNameBinary == null) {
+        String algorithmName = getAlgorithmName();
+        if (algorithmName == null || algorithmName.length() == 0) {
+          throw new IllegalArgumentException("algorithmName cannot be null or empty.");
+        }
+
+        byte[] binary = algorithmName.getBytes(StandardCharsets.UTF_8);
+        if (binary.length > MAX_ALGORITHM_NAME_LENGTH) {
+          throw new IllegalArgumentException("algorithmName " + algorithmName + " exceeds the size of " +
+                  MAX_ALGORITHM_NAME_LENGTH + " bytes.");
+        }
+
+        byte[] binaryWithSize = new byte[binary.length + 1];
+        binaryWithSize[0] = (byte) binary.length;
+        System.arraycopy(binary, 0, binaryWithSize, 1, binary.length);
+
+        algorithmNameBinary = binaryWithSize;
+      }
+    }
+
+    return algorithmNameBinary;
+  }
+
+  /**
+   * Given the source data size, what is the maximum compressed buffer size?
+   * The max buffer size must sufficient to hold:
+   * - version (1 byte)
+   * - algorithm name binary.
+   * - original data size (4 bytes)
+   * - worst-case compression output.
+   * This is an estimate calculation it is usually slightly larger than the sourceDataSize.
+   *
+   * @param sourceDataSize The source data size.
+   * @return The estimated buffer size required to hold the compressed data.
+   */
+  public int getCompressBufferSize(int sourceDataSize) {
+    return estimateMaxCompressedDataSize(sourceDataSize) + getAlgorithmNameBinary().length + 5;
```

Comment:
The size could also depend on the version. A newer version might introduce a new header field.